vault/identity_store_oidc_provider.go

```diff
@@ -1504,20 +1588,473 @@ func authResponse(code, state, errorCode, errorDescription string) (*logical.Res
+		"expires_in":   int64(accessTokenExpiry.Sub(accessTokenIssuedAt).Seconds()),
+	}, "", "")
+}
+
+// tokenResponse returns the OIDC Token Response. An error response is
+// returned if the given error code is non-empty. For details, see spec at
+//   - https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse
+//   - https://openid.net/specs/openid-connect-core-1_0.html#TokenErrorResponse
+func tokenResponse(response map[string]interface{}, errorCode, errorDescription string) (*logical.Response, error) {
+	statusCode := http.StatusOK
+
+	// Set the error response and status code if error code isn't empty
+	if errorCode != "" {
+		statusCode = http.StatusBadRequest
+		response = map[string]interface{}{
+			"error":             errorCode,
+			"error_description": errorDescription,
+		}
+	}
+
+	body, err := json.Marshal(response)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			logical.HTTPStatusCode:  statusCode,
+			logical.HTTPRawBody:     body,
+			logical.HTTPContentType: "application/json",
+
+			// Token responses must include the following HTTP response headers
+			// https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse
+			logical.HTTPCacheControlHeader: "no-store",
+			logical.HTTPPragmaHeader:       "no-cache",
+		},
+	}, nil
+}
+
+func (i *IdentityStore) pathOIDCUserInfo(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	// Get the namespace
+	ns, err := namespace.FromContext(ctx)
+	if err != nil {
+		return userInfoResponse(nil, ErrUserInfoServerError, err.Error())
+	}
+
+	// Get the OIDC provider
+	name := d.Get("name").(string)
+	provider, err := i.getOIDCProvider(ctx, req.Storage, name)
+	if err != nil {
+		return userInfoResponse(nil, ErrUserInfoServerError, err.Error())
+	}
+	if provider == nil {
+		return userInfoResponse(nil, ErrUserInfoInvalidRequest, "provider not found")
+	}
+
+	// Validate that the access token was sent as a Bearer token
+	if req.ClientTokenSource != logical.ClientTokenFromAuthzHeader {
+		return userInfoResponse(nil, ErrUserInfoInvalidToken, "access token must be sent as a Bearer token")
+	}
+
+	// Look up the access token
+	te, err := i.tokenStorer.TokenStore().Lookup(ctx, req.ClientToken)
+	if err != nil {
+		return userInfoResponse(nil, ErrUserInfoServerError, err.Error())
+	}
+	if te == nil {
+		return userInfoResponse(nil, ErrUserInfoInvalidToken, "access token is expired, revoked, malformed, or invalid")
+	}
+	if te.Type != logical.TokenTypeBatch {
+		return userInfoResponse(nil, ErrUserInfoInvalidToken, "access token is expired, revoked, malformed, or invalid")
+	}
+
+	// Get the client ID that originated the request from the token metadata
+	clientID, ok := te.Meta[accessTokenClientIDMeta]
+	if !ok {
+		return userInfoResponse(nil, ErrUserInfoServerError, "expected client ID in token metadata")
+	}
+	client, err := i.clientByID(ctx, req.Storage, clientID)
```

Comment:
will there be any relationship between oauth clients and identity token roles defined under `identity/oidc/token/:name`/`identity/oidc/role/:name`?

In the OIDC provider flow plugin I mentioned in my other client, we use the `client_id` as the `name` in a call to `identity/oidc/token/:name` and configure all of our `identity/oidc/role/:name` configurations so that the [client_id](https://www.vaultproject.io/api/secret/identity/tokens#client_id) on the `identity/oidc/role/:name` configuration is the same as the `name` parameter, which ensure that the returned token has the client_id the resource server provided included in the `aud` claim of the returned token as required by the [oidc spec's id token definition](https://openid.net/specs/openid-connect-core-1_0.html#IDToken)

Reply:
👍 sounds good

It looks like the OIDC clients based tokens support the same templating the existing identity token system has [identity_store_oidc_provider.go#L1944](https://github.com/hashicorp/vault/blob/3c3d4a678b679ab88846cf0aefd413a180ec731e/vault/identity_store_oidc_provider.go#L1944) so we I think we'll be able to template in some metadata and groups membership claims there, which is great,

I think we should be able to make both the OIDC flow issued identity tokens for a client and identity tokens issued for an existing `identity/oidc/token/:name` contain all the same claims, so we should be able to use continue using both interchangeably, which is great for some use cases where we have machines retrieving identity tokens directly to use (effectively) as [SPIFFE JWT's](https://github.com/spiffe/spiffe/blob/main/standards/JWT-SVID.md) for authenticating to internal things, which I'm not sure we'd want to replace with OIDC flows

🙏 thanks again for all the info and replies